pubtools/pulplib/_impl/fake/client.py

```diff
@@ -1,18 +1,20 @@
 import random
 import uuid
 import threading
+import hashlib
 
 from collections import namedtuple
 
 import six
-from more_executors.futures import f_return, f_return_error
+from more_executors.futures import f_return, f_return_error, f_flat_map
 
 from pubtools.pulplib import Page, PulpException, Criteria, Task
 from .. import compat_attr as attr
 
 from .match import match_object
 
 Publish = namedtuple("Publish", ["repository", "tasks"])
+Upload = namedtuple("Upload", ["repository", "tasks", "unit_type_id", "unit_key"])
```

Comment:
I think this is not too good either, exposing the unit_type_id and unit_key here encourages devs to write tests against Pulp2 implementation details where it's not necessary to do so.  upload_file is uploading a file "somehow" and the specifics of unit_type_id, unit_key are intentionally hidden, it's undermined by exposing them here like this.

If we are adding more attributes here I think they should try to be generic so as to expose as few Pulp2 implementation details as possible, while still meeting the requirements.  So, what requirement is being met by these?

I think it's probably: allowing the developer of a test to verify that correct content was uploaded to correct path.  Then I think it should be done more directly, like exposing "name" and ("sha256" or "content") attributes.

If you expose unit_key here, technically developers can look at unit_key["checksum"] to get the checksum.  But that kind of implementation detail - "On Pulp, ISO units have a 'checksum' in unit key" - is exactly what this library is meant to be hiding.